File: web05/src/main/webapp/hr/ember/js/controllers/employee_controller.js
'use strict';

App.EmployeeController = Ember.ObjectController.extend({
	no: null,
	name: null,
	job: null,
	managerNo: null,
	hireDate: null,
	salary: null,
	commission: null,
	departmentNo: null,
	
	empty: function() {
		this.set('no', null);
		this.set('name', null);
		this.set('job', null);
		this.set('managerNo', null);
		this.set('hireDate', null);
		this.set('salary', null);
		this.set('commission', null);
		this.set('departmentNo', null);
	}, 
	
	employee: function(key, value) {
		if (arguments.length == 1) {
			return this.get('no');
		} else {
			var no = this.get('no'); 
			if (value != no) {
				this.empty();
				
				var controller = this;
				jQuery.getJSON('../retrieveEmp.do?empno=' + value, function(result){
					if (result.status == "success") {
						var emp = result.data;
						controller.set('no', emp.no);
						controller.set('name', emp.name);
						controller.set('job', emp.job);
						controller.set('managerNo', emp.managerNo);
						controller.set('hireDate', emp.hireDate);
						controller.set('salary', emp.salary);
						controller.set('commission', emp.commission);
						controller.set('departmentNo', emp.departmentNo);
					} else {
						controller.transitionToRoute('employees', 1);
					}
				});
			}
		}
	}.property(),
	
	add: function() {
		var controller = this;
		$.post('../addEmp.do',
			{
				no: this.get('no'),
				name: this.get('name'),
				job: this.get('job'),
				managerNo: this.get('managerNo'),
				hireDate: this.get('hireDate'),
				salary: this.get('salary'),
				commission: this.get('commission'),
				departmentNo: this.get('departmentNo')
			},
			function(result) {
				if (result.status == 'success') {
					controller.transitionToRoute('employees', 1);
				} else {
					alert(result.message);
				}
			},
			'json'
		);
	},
	
	update: function() {
		var controller = this;
		$.post('../updateEmp.do', 
			{
				no: this.get('no'),
				name: this.get('name'),
				job: this.get('job'),
				managerNo: this.get('managerNo'),
				hireDate: this.get('hireDate'),
				salary: this.get('salary'),
				commission: this.get('commission'),
				departmentNo: this.get('departmentNo')
			}, 
		    function(result) {
				if (result.status == 'success') {
					controller.transitionToRoute('employees');
				} else {
					alert(result.message);
				}
			}, 
			'json'
		);
	},

	remove: function() {
		var controller = this;
		$.getJSON('../deleteEmp.do?empno=' + this.get('no'), 
			function(result) {
				if (result.status == 'success') {
					controller.transitionToRoute('employees');
				} else {
					alert(result.message);
				}
			}
		);
	},

});

File: web05/src/main/java/bit/java39/controls/hr/EmployeeController.java
package bit.java39.controls.hr;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.propertyeditors.CustomDateEditor;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.InitBinder;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;


import bit.java39.service.EmployeeService;
import bit.java39.vo.Employee;

@Controller
@RequestMapping("/hr")
public class EmployeeController {
	@Autowired
	EmployeeService employeeService;
	
	@InitBinder
	public void initBinder(WebDataBinder binder) {
		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
		dateFormat.setLenient(false);
		binder.registerCustomEditor(Date.class, new CustomDateEditor(dateFormat, false));
	}

	/*
	@RequestMapping("/searchEmp")
	public ResponseEntity<String> list(
			@RequestParam(defaultValue="0") int deptno) throws Exception {
		
		HttpHeaders headers = new HttpHeaders();
		headers.add("Test", "okok");
		
		return new ResponseEntity<String>(
				new Gson().toJson( employeeDao.getEmployeeList(deptno) ),
				headers,
				HttpStatus.CREATED);
	}
	*/
	
	/*
	@RequestMapping("/searchEmp")
	@ResponseBody
	public String list(
			@RequestParam(defaultValue="0") int deptno) throws Exception {
		return new Gson().toJson( employeeDao.getEmployeeList(deptno) );
	}
	*/
	
	/*
	@RequestMapping(value="/searchEmp")
	public ModelAndView listForJSP(
			@RequestParam(defaultValue="0") int deptno) throws Exception {
		
		ModelAndView mv = new ModelAndView();
		mv.addObject("employeeList", employeeDao.getEmployeeList(deptno));
		mv.setViewName("hr/employeeList");
		return mv;
	}
	*/
	
	/* Object => JSON 출력
	 * - Object를 리턴할 때는 JSON으로 변환하는 ViewResolver에게 전달하도록 설정해야 한다.
	 * - @ResponseBody String을 리턴할 때는 직접 출력하는 ViewResolver에게 전달하도록 설정.
	 * - 일반 String일 경우는 JSP로 포워딩 하는 ViewResolver에게 전달하도록 설정한다.
	 * - 이것을 중간에서 중재해주는 ViewResolver가 있다. 
	 */
	@RequestMapping(value="/searchEmp")
	@ResponseBody
	public Object list(
			@RequestParam(defaultValue="1") int pageNo,
			@RequestParam(defaultValue="5") int pageSize) throws Exception {
		HashMap<String,Object> responseResult = new HashMap<String,Object>();
		try {
			responseResult.put("employeeList", employeeService.getEmployeeList(pageNo, pageSize));
			responseResult.put("status", "success");
			responseResult.put("pageNo", pageNo);
			responseResult.put("pageSize", pageSize);
			
			int empCount = employeeService.countAll();
			int totalPage =  empCount / pageSize;
			if (empCount % pageSize > 0) {
				totalPage++; 
			}
			responseResult.put("totalPage", totalPage);
			
		} catch (Exception e) {
			responseResult.put("status", "fail");
			
			StringWriter temp = new StringWriter();
			PrintWriter out = new PrintWriter(temp);
			e.printStackTrace(out);
			
			responseResult.put("message", temp.toString());
		}
		return responseResult;
		
		
		
	}
	
	@RequestMapping(value="/addEmp",method=RequestMethod.POST)
	@ResponseBody
	public HashMap<String,Object> add(Employee employee, BindingResult result) throws Exception {
		HashMap<String,Object> responseResult = new HashMap<String,Object>();
		try {
			int count = employeeService.addEmployee(employee);
			
			if (count > 0) {
				responseResult.put("status", "success");
			} else {
				responseResult.put("status", "fail");
				responseResult.put("message", "Insert count is 0.");
			}
		} catch (Exception e) {
			responseResult.put("status", "fail");
			
			StringWriter temp = new StringWriter();
			PrintWriter out = new PrintWriter(temp);
			e.printStackTrace(out);
			
			responseResult.put("message", temp.toString());
		}
		return responseResult;
	}
	
	@RequestMapping("/retrieveEmp")
	@ResponseBody
	public Object retrieve(
			@RequestParam("empno") int empno, 
			ModelMap modelMap) throws Exception {
		HashMap<String,Object> responseResult = new HashMap<String,Object>();
		try {
			Employee emp = employeeService.getEmployee(empno);
			if (emp != null) {
				responseResult.put("status", "success");
				responseResult.put("data", emp);
			} else {
				responseResult.put("status", "fail");
				responseResult.put("message", "해당 직원이 없습니다!");
			}
		} catch (Exception e) {
			responseResult.put("status", "fail");
			
			StringWriter temp = new StringWriter();
			PrintWriter out = new PrintWriter(temp);
			e.printStackTrace(out);
			
			responseResult.put("message", temp.toString());
		}
		
		return responseResult;
	}

	@RequestMapping(value="/updateEmp", method=RequestMethod.POST)
	@ResponseBody
	public Object update(Employee employee, BindingResult result) throws Exception {
		HashMap<String,Object> responseResult = new HashMap<String,Object>();
		try {
			int count = employeeService.changeEmployee(employee);
		
			if (count > 0) {
				responseResult.put("status", "success");
			} else {
				responseResult.put("status", "fail");
				responseResult.put("message", "해당 직원이 없습니다!");
			}
		} catch (Exception e) {
			responseResult.put("status", "fail");
			
			StringWriter temp = new StringWriter();
			PrintWriter out = new PrintWriter(temp);
			e.printStackTrace(out);
			
			responseResult.put("message", temp.toString());
		}
		
		return responseResult;
	}
	
	@RequestMapping("/deleteEmp")
	@ResponseBody
	public Object delete(
			@RequestParam(value="empno",defaultValue="0") int empno) throws Exception {
		HashMap<String,Object> responseResult = new HashMap<String,Object>();
		try {
			int count = employeeService.removeEmployee(empno);
		
			if (count > 0) {
				responseResult.put("status", "success");
			} else {
				responseResult.put("status", "fail");
				responseResult.put("message", "해당 직원이 없습니다!");
			}
		} catch (Exception e) {
			responseResult.put("status", "fail");
			
			StringWriter temp = new StringWriter();
			PrintWriter out = new PrintWriter(temp);
			e.printStackTrace(out);
			
			responseResult.put("message", temp.toString());
		}
		
		return responseResult;
	}
	
}









